### PR TITLE
Improve ICE message for forbidden dep-graph reads.

### PR DESCRIPTION
**Pull request by michaelwoerister**
The new message mentions the main context that the ICE might occur in and it mentions the query/dep-node that is being read.

cc https://github.com/rust-lang/rust/pull/123781, where this would have been helpful.

### CODE REVIEW

**Comment by oli-obk** [comment]
That's a significantly more useful message than I expected 😆 

@bors r+ rollup

r? @oli-obk

File: compiler/rustc_query_system/src/dep_graph/graph.rs [hunk 1]
```diff
@@ -1366,3 +1367,41 @@ pub(crate) fn print_markframe_trace<D: Deps>(graph: &DepGraph<D>, frame: Option<
...
+    }
+
+    let dep_node = dep_node.map_or_else(
+        || format!("with index {:?}", dep_node_index),
+        |dep_node| format!("`{:?}`", dep_node),
+    );
+
+    panic!(
+        "Error: trying to record dependency on DepNode {dep_node} in a \
+         context that does not allow it (e.g. during query deserialization)."
```
Comment (RalfJung) [inline compiler/rustc_query_system/src/dep_graph/graph.rs:1405]: FWIW, as someone who knows nothing about the query/incremental system, I still don't know that "record dependency" is the same as "invoke query". Or maybe it's not the same, but the explanation you gave me that I actually understood was "you invoked a query in a context that does not allow it", and *that* I can understand. But maybe there are cases where saying "invoked a query" here would be wrong?

Comment (michaelwoerister) [reply compiler/rustc_query_system/src/dep_graph/graph.rs]: Is the new message better?

File: compiler/rustc_query_system/src/dep_graph/graph.rs [hunk 2]
```diff
@@ -1402,6 +1402,10 @@ fn panic_on_forbidden_read<D: Deps>(data: &DepGraphData<D>, dep_node_index: DepN
 
     panic!(
         "Error: trying to record dependency on DepNode {dep_node} in a \
-         context that does not allow it (e.g. during query deserialization)."
+         context that does not allow it (e.g. during query deserialization). \
+         The most common case of recording a dependency on a DepNode `foo` is \
+         when the correspondng query `foo` is invoked. Invoking queries is not \
+         allowed as part of loading something from the incremental on-disk cache. \
+         See <https://github.com/rust-lang/rust/pull/91919>."
```
Comment (RalfJung) [inline compiler/rustc_query_system/src/dep_graph/graph.rs:1409]: That's a lot better, thanks!

Not sure if linking to the PR is very useful, though.

Comment (michaelwoerister) [reply compiler/rustc_query_system/src/dep_graph/graph.rs:1409]: I'll leave the link in unless you or @oli-obk think that it's actively hurting understandability.

**Comment by michaelwoerister** [comment]
Updated the error message: https://github.com/rust-lang/rust/pull/124252/commits/6146a51f17f19b557c2baf11a1ae04cbafdd89bb

@bors r=oli-obk

**Pull request merged**